gcc build issue with CHECK_* macros

I get a 4 page long screen of error messages which contains:

```
In file included from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:31:0,
                 from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.cpp:27:
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [11]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_con
vertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[11]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [11]]'
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:197:16:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
         if (value == nullptr)
             ~~~~~~^~~~~~~~~~
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [13]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_con
vertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[13]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [13]]'
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:199:16:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [16]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_con
vertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[16]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [16]]'
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:134:24:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [5]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_conv
ertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[5]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [5]]'
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:137:24:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [15]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_con
vertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[15]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [15]]'
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:137:41:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In instantiation of 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [7]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_
...
value && (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [2]; std::stringstream = std::__cxx11::basic_stringstream<char>; typename std::enable_if<(std::is_conv
ertible<T, const char*>::value && (! std::is_null_pointer<_Tp>::value)), const T&>::type = const char (&)[2]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:108:25:   required from 'Logging::CheckException& Logging::CheckException::operator<<(const T&) [with T = char [2]]'
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:160:35:   required from 'Logging::CheckException Logging::CheckResult<T1, T2>::Exception(const char*, const char*, const char*) 
const [with T1 = BitFunnel::ByteCodeInterpreter::Opcode; T2 = BitFunnel::ByteCodeInterpreter::Opcode]'
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:133:17:   required from here
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:19: error: the compiler can assume that the address of 'value' will never be NULL [-Werror=address]
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [13]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
         ^~
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [16]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
         ^~
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [5]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
         ^~
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [15]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
         ^~
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [7]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
...
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h: In function 'void Logging::StreamOut(std::stringstream&, typename std::enable_if<(std::is_convertible<T, const char*>::value &&
 (! std::is_null_pointer<_Tp>::value)), const T&>::type) [with T = char [2]]':
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:74:9: error: nonnull argument 'value' compared to NULL [-Werror=nonnull-compare]
         if (value == nullptr)
         ^~
In file included from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.cpp:27:0:
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h: In constructor 'BitFunnel::ByteCodeInterpreter::Instruction::Instruction(BitFunnel::ByteCodeInterpreter::Opcode, size_t, 
size_t, bool)':
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:220:43: error: '<anonymous>' may be used uninitialized in this function [-Werror=maybe-uninitialized]
         out << static_cast<unsigned>(value);
                                           ^
In file included from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:31:0,
                 from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.cpp:27:
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:252:45: note: '<anonymous>' was declared here
     ::Logging::CmpHelper##op_name(val1, val2))       \
                                             ^
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:265:5: note: in expansion of macro 'CHECK_OP'
     CHECK_OP(val1, val2, LT)
     ^~~~~~~~
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:133:17: note: in expansion of macro 'CHECK_LT'
                 CHECK_LT(opcode, Opcode::Last)
                 ^~~~~~~~
```

I'll probably take a look at this later tonight or sometime tomorrow.

After 0fa6fe6, we're down to this, much shorter, error:

~~~
In file included from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.cpp:27:0:
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h: In constructor 'BitFunnel::ByteCodeInterpreter::Instruction::Instruction(BitFunnel::ByteCodeInterpreter::Opcode, size_t, size_t, bool)':
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:220:43: error: '<anonymous>' may be used uninitialized in this function [-Werror=maybe-uninitialized]
         out << static_cast<unsigned>(value);
                                           ^
In file included from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:31:0,
                 from /home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.cpp:27:
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:246:45: note: '<anonymous>' was declared here
     ::Logging::CmpHelper##op_name(val1, val2))       \
                                             ^
/home/leah/dev/BitFunnel/inc/LoggerInterfaces/Check.h:259:5: note: in expansion of macro 'CHECK_OP'
     CHECK_OP(val1, val2, LT)
     ^~~~~~~~
/home/leah/dev/BitFunnel/src/Plan/src/ByteCodeInterpreter.h:133:17: note: in expansion of macro 'CHECK_LT'
                 CHECK_LT(opcode, Opcode::Last)
                 ^~~~~~~~
~~~

--- a/inc/LoggerInterfaces/Check.h
+++ b/inc/LoggerInterfaces/Check.h
@@ -47,7 +47,16 @@ namespace Logging
                                            !std::is_pointer<T>::value &&
                                            !std::is_convertible<T, char const *>::value, T const &>::type value)
     {
+        // Work-around for what is believed to be a bug in g++ v5.4.0 (among
+        // other versions). See detailed discussion in #225.
+#ifndef BITFUNNEL_PLATFORM_WINDOWS
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
         stream << value;
+#ifndef BITFUNNEL_PLATFORM_WINDOWS
+#pragma GCC diagnostic pop
+#endif
     }
 
     template <typename T>
@@ -56,7 +65,16 @@ namespace Logging
                             !std::is_convertible<T, char const *>::value) ||
                             std::is_null_pointer<T>::value, T const &>::type value)
     {
+        // Work-around for what is believed to be a bug in g++ v5.4.0 (among
+        // other versions). See detailed discussion in #225.
+#ifndef BITFUNNEL_PLATFORM_WINDOWS
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
         if (value == nullptr)
+#ifndef BITFUNNEL_PLATFORM_WINDOWS
+#pragma GCC diagnostic pop
+#endif
         {
             stream << "(nullptr)";
         }

--- a/src/Plan/src/ByteCodeInterpreter.h
+++ b/src/Plan/src/ByteCodeInterpreter.h
@@ -257,7 +257,16 @@ namespace BitFunnel
 
     inline std::ostream& operator<<(std::ostream& out, const ByteCodeInterpreter::Opcode value)
     {
+        // Work-around for what is believed to be a bug in g++ v5.4.0 (among
+        // other versions). See detailed discussion in #225.
+#ifndef BITFUNNEL_PLATFORM_WINDOWS
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
         out << c_opcodeNames[static_cast<size_t>(value)];
+#ifndef BITFUNNEL_PLATFORM_WINDOWS
+#pragma GCC diagnostic pop
+#endif
         return out;
     }
 

